### ISSUE TITLE
Relax -Wreorder-init-list for POD datatypes

### ISSUE THREAD

**Issue by pirama-arumuga-nainar** [llvm/llvm-project#81643]
-Wreorder-init-list warns if fields in an initializer list are set out of order.  The potential bugs are always related to initialization order of non-POD fields.

There are numerous instances of this warning in the Android codebase and majority are for POD types - drowning the few cases that need to be fixed.  Secondly, due to historical reason, one of these POD structs has different order of fields for 32-vs-64-bit so it's infeasible to fix these warnings in a reasonable fashion.
 
To make this warning more useful for Android (and maybe other platforms), can we relax this warning for POD datatypes?  Another option is to possibly create a sub-warning that only warns on POD types that can be disabled in Android's build system?

(cc: @enh-google @ZijunZhaoCCK @AaronBallman )

**Comment by pinskia** [comment]
I hear that the Linux kernel folks would like to be able to the ability to disable the warning just for "POD" structures too.
This comes from the GCC bug https://gcc.gnu.org/bugzilla/show_bug.cgi?id=113124 where GCC currently rejects all reordering.

**Comment by enh-google** [comment]
> due to historical reason

for a bit of background on this, lest this read as "you can just clean up your code, no?" --- POSIX tells you the names of fields that will be in structs, but it doesn't tell you (a) all of them nor (b) what _order_ they will be in. so there are plenty of POSIX structs (in libc headers) where there is no portable "correct" order, because different operating systems disagree. but even if you only care about Linux, say, there are still structs where Linux disagrees with itself on the ordering between different architectures.

`struct sigaction` is the textbook example of "if my code uses this struct, i have to turn this warning off". which does seem unfortunate, because it seems like there might be _some_ value to this warning if non-POD types [or whatever they're called these days] are involved?

i think @pinskia is telling us that C++20 explicitly specifies that they're required to be in order [but you can still have missing ones?]. /me struggles to believe that no-one on the C++20 committee thought about interop with C libraries... i don't suppose there was a defect report for C++23? is there one for C++26?

**Comment by erichkeane** [comment]
> i think @pinskia is telling us that C++20 explicitly specifies that they're required to be in order [but you can still have missing ones?]. /me struggles to believe that no-one on the C++20 committee thought about interop with C libraries... i don't suppose there was a defect report for C++23? is there one for C++26?

Sadly this was discussed in the C++ commitee, and the C interop was decided as not particularly relevant.  There was a belief that the 'init vs specified' order  (in that C++ essentially mandates the init order must be lexically in the struct) would be too confusing to allow this.

That said, "POD" doesn't have meaning anymore, we've actually split it into 'trivial' and 'standard layout'.   FOR the purposes of construction via aggregate initialization with designated init, I think the restrictions on class-type aggregate init are more than sufficient to ignore this warning if you wish: I don't see a way of deciding only SOME of the aggregate rules deserve an additional relaxation/subset of the rule.  

I could PERHAPS be convinced 'default member init' would cause the more severe warning, or perhaps even base classes, but something like a destructor's presence wouldn't be motivating to me.

So in summary: I'm sympathetic, but I don't see how such a relaxation would make sense in a case where you wouldn't just disable this diagnostic anyway.
